core/clustersmngr/factory.go

```diff
@@ -0,0 +1,306 @@
+			if err := c.List(ctx, nsList); err != nil {
+				cf.log.Error(err, "failed listing namespaces", "cluster", clusterName)
+			}
+
+			cf.clustersNamespaces.Set(clusterName, nsList.Items)
+		}(clusterName, c)
+	}
+
+	wg.Wait()
+
+	return nil
+}
+
+func (cf *clientsFactory) GetImpersonatedClient(ctx context.Context, user *auth.UserPrincipal) (Client, error) {
+	pool := NewClustersClientsPool()
+
+	for _, cluster := range cf.clusters.Get() {
+		if err := pool.Add(ClientConfigWithUser(user), cluster); err != nil {
+			return nil, fmt.Errorf("failed adding cluster client to pool: %w", err)
+		}
+	}
+
+	return NewClient(pool, cf.userNsList(ctx, user)), nil
+}
+
+func (cf *clientsFactory) GetServerClient(ctx context.Context) (Client, error) {
+	pool := NewClustersClientsPool()
+
+	for _, cluster := range cf.clusters.Get() {
+		if err := pool.Add(restConfigFromCluster, cluster); err != nil {
+			return nil, fmt.Errorf("failed adding cluster client to pool: %w", err)
+		}
+	}
+
+	return NewClient(pool, cf.clustersNamespaces.namespaces), nil
+}
+
+func (cf *clientsFactory) GetClustersNamespaces() map[string][]v1.Namespace {
+	return cf.clustersNamespaces.namespaces
+}
+
+func restConfigFromCluster(cluster Cluster) *rest.Config {
+	return &rest.Config{
+		Host:            cluster.Server,
+		BearerToken:     cluster.BearerToken,
+		TLSClientConfig: cluster.TLSConfig,
+		QPS:             ClientQPS,
+		Burst:           ClientBurst,
+	}
+}
+
+func (cf *clientsFactory) userNsList(ctx context.Context, user *auth.UserPrincipal) map[string][]v1.Namespace {
+	userNamespaces := cf.usersNamespaces.GetAll(user, cf.clusters.Get())
+	if len(userNamespaces) > 0 {
+		return userNamespaces
+	}
+
+	wg := sync.WaitGroup{}
+
+	for _, cluster := range cf.clusters.Get() {
+		wg.Add(1)
+
+		go func(cluster Cluster) {
+			defer wg.Done()
+
+			clusterNs := cf.clustersNamespaces.Get(cluster.Name)
+
+			filteredNs, err := cf.nsChecker.FilterAccessibleNamespaces(ctx, impersonatedConfig(cluster, user), clusterNs)
+			if err != nil {
+				cf.log.Error(err, "failed filtering namespaces", "cluster", cluster.Name, "user", user.ID)
+			}
+
+			cf.usersNamespaces.Set(user, cluster.Name, filteredNs)
+		}(cluster)
+	}
+
+	wg.Wait()
+
+	return cf.usersNamespaces.GetAll(user, cf.clusters.Get())
+}
+
+func impersonatedConfig(cluster Cluster, user *auth.UserPrincipal) *rest.Config {
+	shallowCopy := *restConfigFromCluster(cluster)
+
+	shallowCopy.Impersonate = rest.ImpersonationConfig{
+		UserName: user.ID,
+		Groups:   user.Groups,
+	}
+
+	return &shallowCopy
+}
+
+func clientsForClusters(clusters []Cluster) (map[string]client.Client, error) {
+	clients := map[string]client.Client{}
+
+	for _, cluster := range clusters {
+		c, err := client.New(restConfigFromCluster(cluster), client.Options{})
+		if err != nil {
+			return nil, fmt.Errorf("failed creating client for cluster %s: %w", cluster.Name, err)
+		}
+
+		clients[cluster.Name] = c
+	}
+
+	return clients, nil
+}
+
+type Clusters struct {
```

Comment:
nit: can we move everything from here down into like `x_cache` files? not bothered if you'd rather not, just a preference for keeping things tidy and organised away from the main "drama" of the file

Reply:
fair enough